src/main/java/com/pingcap/tikv/PDClient.java

```diff
@@ -229,60 +224,65 @@ void close() {
-    String leaderUrlStr = "URL Not Set";
-    try {
-      long ts = System.nanoTime();
-      synchronized (this) {
-        // Lock for not flooding during pd error
-        if (leaderWrapper != null && leaderWrapper.getCreateTime() > ts) return;
-
-        if (resp == null) {
-          resp = getMembers();
-          if (resp == null) return;
-        }
-        Member leader = resp.getLeader();
-        List<String> leaderUrls = leader.getClientUrlsList();
-        if (leaderUrls.isEmpty()) return;
-        leaderUrlStr = leaderUrls.get(0);
-        URL tURL = new URL(leaderUrlStr);
-        HostAndPort newLeader = HostAndPort.fromParts(tURL.getHost(), tURL.getPort());
-        leaderUrlStr = newLeader.toString();
-        // TODO: Why not strip protocol info on server side since grpc does not need it
-        if (leaderWrapper != null && leaderUrlStr.equals(leaderWrapper.getLeaderInfo())) {
-          return;
-        }
+  private void switchLeader(List<String> leaderURLs) {
+    if(leaderURLs.isEmpty()) return;
+    String leaderUrlStr = leaderURLs.get(0);
+    // TODO: Why not strip protocol info on server side since grpc does not need it
+    if (leaderWrapper != null && leaderUrlStr.equals(leaderWrapper.getLeaderInfo())) {
+      return;
+    }
+    // switch leader
+    createLeaderWrapper(leaderUrlStr);
+   logger.info(String.format("Switched to new leader: %s", leaderWrapper));
+  }
 
-        // switch leader
-        ManagedChannel clientChannel = getChannel(leaderUrlStr);
-        leaderWrapper =
-            new LeaderWrapper(
-                leaderUrlStr,
-                PDGrpc.newBlockingStub(clientChannel),
-                PDGrpc.newStub(clientChannel),
-                System.nanoTime());
-        logger.info(String.format("Switched to new leader: %s", leaderWrapper));
+  private void createLeaderWrapper(String leaderUrlStr) {
+    try {
+      URL tURL = new URL(leaderUrlStr);
+      HostAndPort newLeader = HostAndPort.fromParts(tURL.getHost(), tURL.getPort());
+      leaderUrlStr = newLeader.toString();
+      if (leaderWrapper != null && leaderUrlStr.equals(leaderWrapper.getLeaderInfo())) {
+        return;
       }
-    } catch (Exception e) {
+
+      // create new Leader
+      ManagedChannel clientChannel = getChannel(leaderUrlStr);
+      leaderWrapper =
+        new LeaderWrapper(
+            leaderUrlStr,
+            PDGrpc.newBlockingStub(clientChannel),
+            PDGrpc.newStub(clientChannel),
+            System.nanoTime());
+    } catch (MalformedURLException e) {
       logger.error("Error updating leader.", e);
     }
   }
 
+  public void updateLeader() {
+    for(HostAndPort url : this.pdAddrs) {
+      // since resp is null, we need update leader's address by walking through all pd server.
+      GetMembersResponse resp = getMembers(url);
+      if(resp == null) {
+        continue;
+      }
+      switchLeader(resp.getLeader().getClientUrlsList());
```

Comment:
where is return? Isn't this code throw anyway?

Reply:
FIXED.